# MATH 210 Introduction to Mathematical Computing

## March 28, 2018

* import with pandas.read_csv

In [1]:
import numpy as np
import matplotlib.pyplot as plt
import scipy.integrate as spi
import pandas as pd
%matplotlib inline

### Importing data with pandas.read_csv

In [2]:
data = pd.read_csv('04770123-eng.csv')

UnicodeDecodeError: 'utf-8' codec can't decode byte 0xe9 in position 9: invalid continuation byte

There is an encording error. This dataset comes from the Goverment of Canada and there are probably some French accent characters that pandas doesn't recognize. Let's try again and specify `encoding='Latin'` to pandas that there are accent characters in the data.

In [4]:
data = pd.read_csv('04770123-eng.csv',encoding='latin')

The data is imported as a panda DataFrame. This is like a spreadsheet with column names and row index. To get a summary of the data, let's get a summary of the data, let's use the function `.info()`.

In [5]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5159 entries, 0 to 5158
Data columns (total 8 columns):
Ref_Date      5159 non-null object
GEO           5159 non-null object
INS           5159 non-null object
RAN           5159 non-null object
STA           5159 non-null object
Vector        5159 non-null object
Coordinate    5159 non-null object
Value         5159 non-null object
dtypes: object(8)
memory usage: 322.5+ KB


We have 5159 rows of data and 8 columns. The word `object` menas text. That's fine for most of the columns since `INS` is the name of the university but `value` should be a number. Let's use the function `.head()` to see the first 5 rows of the data.

In [6]:
data.head()

Ref_Date     GEO                                                INS  \
0  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
1  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
2  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
3  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
4  2016/2017  Canada  Memorial University of Newfoundland - Includin...   

              RAN                        STA      Vector Coordinate     Value  
0  Full professor       Total teaching staff  v113286602    1.1.2.1     294.0  
1  Full professor    Teaching staff excluded  v113286603    1.1.2.2      12.0  
2  Full professor          Average (dollars)  v113286604    1.1.2.3  153550.0  
3  Full professor           Median (dollars)  v113286605    1.1.2.4  151325.0  
4  Full professor  10th Percentile (dollars)  v113286606    1.1.2.5  129900.0

Why is the column `Value` imported as text and not a float? Let's take a look. We can select a column using a brackets and clumn name.

In [7]:
data['Value']

0          294.0
1           12.0
2       153550.0
3       151325.0
4       129900.0
5       174875.0
6           42.0
7           12.0
8       159675.0
9       156425.0
10      132825.0
11      188475.0
12         252.0
13      152525.0
14      151325.0
15      129900.0
16      174875.0
17         420.0
18           6.0
19      123500.0
20      120125.0
21      108475.0
22      142750.0
23          60.0
24             x
25      130300.0
26      128750.0
27      114625.0
28      146775.0
29         363.0
          ...   
5129     88300.0
5130        54.0
5131           x
5132     82275.0
5133     85525.0
5134     68500.0
5135     88300.0
5136        12.0
5137     77700.0
5138     77700.0
5139     77700.0
5140     77700.0
5141       255.0
5142         6.0
5143     88050.0
5144     88300.0
5145     77700.0
5146     90325.0
5147       117.0
5148           x
5149     87125.0
5150     88300.0
5151     81725.0
5152     88300.0
5153       138.0
5154           x
5155     88825.0
5156     88300

Missing data in the `Value` column is represented by an `x`. We need to import the data and tell pandas that missing data is called `x`.

In [8]:
data = pd.read_csv('04770123-eng.csv',encoding='latin', na_values = 'x')

In [9]:
data.info

<bound method DataFrame.info of        Ref_Date     GEO                                                INS  \
0     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
1     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
2     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
3     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
4     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
5     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
6     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
7     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
8     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
9     2016/2017  Canada  Memorial University of Newfoundland - Includin...   
10    2016/2017  Canada  Memorial University of Newfoundland - Includin...   
11    2016/2017  Canada  Memoria

In [10]:
data.head()

Ref_Date     GEO                                                INS  \
0  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
1  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
2  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
3  2016/2017  Canada  Memorial University of Newfoundland - Includin...   
4  2016/2017  Canada  Memorial University of Newfoundland - Includin...   

              RAN                        STA      Vector Coordinate     Value  
0  Full professor       Total teaching staff  v113286602    1.1.2.1     294.0  
1  Full professor    Teaching staff excluded  v113286603    1.1.2.2      12.0  
2  Full professor          Average (dollars)  v113286604    1.1.2.3  153550.0  
3  Full professor           Median (dollars)  v113286605    1.1.2.4  151325.0  
4  Full professor  10th Percentile (dollars)  v113286606    1.1.2.5  129900.0

Let's see the different entries in the column `Ref_Date`.

In [11]:
data['Ref_Date'].unique()

array(['2016/2017'], dtype=object)

In [12]:
data['GEO'].unique()

array(['Canada'], dtype=object)

The column `Ref_date` and `GEO` have only one value so let's drop it.

In [13]:
data.drop(['Ref_Date','GEO'], axis = 1, inplace = True)

The columns `Vector` and `Coordinate` 

In [14]:
data['Coordinate'].nunique()

5159

These solumns seem to be IDs. We can drop them too.

In [15]:
data.drop(['Coordinate','Vector'], axis = 1, inplace = True)

Let's take a look.

In [16]:
data.head()

INS             RAN  \
0  Memorial University of Newfoundland - Includin...  Full professor   
1  Memorial University of Newfoundland - Includin...  Full professor   
2  Memorial University of Newfoundland - Includin...  Full professor   
3  Memorial University of Newfoundland - Includin...  Full professor   
4  Memorial University of Newfoundland - Includin...  Full professor   

                         STA     Value  
0       Total teaching staff     294.0  
1    Teaching staff excluded      12.0  
2          Average (dollars)  153550.0  
3           Median (dollars)  151325.0  
4  10th Percentile (dollars)  129900.0

How many different institutions are in the dataset?

In [17]:
data['INS'].unique

<bound method Series.unique of 0       Memorial University of Newfoundland - Includin...
1       Memorial University of Newfoundland - Includin...
2       Memorial University of Newfoundland - Includin...
3       Memorial University of Newfoundland - Includin...
4       Memorial University of Newfoundland - Includin...
5       Memorial University of Newfoundland - Includin...
6       Memorial University of Newfoundland - Includin...
7       Memorial University of Newfoundland - Includin...
8       Memorial University of Newfoundland - Includin...
9       Memorial University of Newfoundland - Includin...
10      Memorial University of Newfoundland - Includin...
11      Memorial University of Newfoundland - Includin...
12      Memorial University of Newfoundland - Includin...
13      Memorial University of Newfoundland - Includin...
14      Memorial University of Newfoundland - Includin...
15      Memorial University of Newfoundland - Includin...
16      Memorial University of Newfoundla

In [18]:
data['INS'].nunique

<bound method IndexOpsMixin.nunique of 0       Memorial University of Newfoundland - Includin...
1       Memorial University of Newfoundland - Includin...
2       Memorial University of Newfoundland - Includin...
3       Memorial University of Newfoundland - Includin...
4       Memorial University of Newfoundland - Includin...
5       Memorial University of Newfoundland - Includin...
6       Memorial University of Newfoundland - Includin...
7       Memorial University of Newfoundland - Includin...
8       Memorial University of Newfoundland - Includin...
9       Memorial University of Newfoundland - Includin...
10      Memorial University of Newfoundland - Includin...
11      Memorial University of Newfoundland - Includin...
12      Memorial University of Newfoundland - Includin...
13      Memorial University of Newfoundland - Includin...
14      Memorial University of Newfoundland - Includin...
15      Memorial University of Newfoundland - Includin...
16      Memorial University of Ne

In [19]:
data['RAN'].unique()

array(['Full professor', 'With senior administrative duties',
       'Without senior administrative duties', 'Associate professor',
       'Assistant professor', 'Rank below assistant professor',
       'All ranks combined (including deans)', 'Male', 'Female',
       'Other teaching staff', 'Visiting Staff', 'Sex unknown/other'], dtype=object)

In [20]:
data['STA'].unique()

array(['Total teaching staff', 'Teaching staff excluded',
       'Average (dollars)', 'Median (dollars)',
       '10th Percentile (dollars)', '90th Percentile (dollars)'], dtype=object)

In [27]:
data_men_women = data[(data['RAN'] == 'Male') | (data['RAN'] == 'Female')]

In [29]:
data_men_women_median = data_men_women[data_men_women['STA'] == 'Median (dollars)']

In [31]:
data_men_women_median.groupby(['INS','RAN'])['Value']